src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Mac.java

```diff
@@ -192,12 +200,53 @@ protected void engineReset() {
     // see JCE spec
     protected void engineInit(Key key, AlgorithmParameterSpec params)
             throws InvalidKeyException, InvalidAlgorithmParameterException {
+        reset(true);
+        p11Key = null;
+        if (svcPbeKi != null) {
+            if (key instanceof P11Key) {
+                // If the key is a P11Key, it must come from a PBE derivation
+                // because this is a PBE Mac service. In addition to checking
+                // the key, check that params (if passed) are consistent.
+                PBEUtil.checkKeyAndParams(key, params, algorithm);
```

Comment:
After this call, shouldn't you store key internally as p11Key?

Reply:
The key is stored as in the internal field p11Key a bit later. The reason is because we need P11SecretKeyFactory::convertKey to check if the P11 key is from a different token and, eventually, re-derive it. The path taken is the same as for non-PBE Mac services. It's only in the PBE Mac case in which we had to derive the PBE key that we have absolute guarantees that calling P11SecretKeyFactory::convertKey would be useless, and we cut it.